File: code/rust/ch07/src/api/lua_state.rs
pub trait LuaState {
    /* basic stack manipulation */
    fn get_top(&self) -> isize;
    fn abs_index(&self, idx: isize) -> isize;
    fn check_stack(&mut self, n: usize) -> bool;
    fn pop(&mut self, n: usize);
    fn copy(&mut self, from_idx: isize, to_idx: isize);
    fn push_value(&mut self, idx: isize);
    fn replace(&mut self, idx: isize);
    fn insert(&mut self, idx: isize);
    fn remove(&mut self, idx: isize);
    fn rotate(&mut self, idx: isize, n: isize);
    fn set_top(&mut self, idx: isize);
    /* access functions (stack -> rust) */
    fn type_name(&self, tp: i8) -> &str; // TODO
    fn type_id(&self, idx: isize) -> i8; // `type` is a keyword
    fn is_none(&self, idx: isize) -> bool;
    fn is_nil(&self, idx: isize) -> bool;
    fn is_none_or_nil(&self, idx: isize) -> bool;
    fn is_boolean(&self, idx: isize) -> bool;
    fn is_integer(&self, idx: isize) -> bool;
    fn is_number(&self, idx: isize) -> bool;
    fn is_string(&self, idx: isize) -> bool;
    fn is_table(&self, idx: isize) -> bool;
    fn is_thread(&self, idx: isize) -> bool;
    fn is_function(&self, idx: isize) -> bool;
    fn to_boolean(&self, idx: isize) -> bool;
    fn to_integer(&self, idx: isize) -> i64;
    fn to_integerx(&self, idx: isize) -> Option<i64>;
    fn to_number(&self, idx: isize) -> f64;
    fn to_numberx(&self, idx: isize) -> Option<f64>;
    fn to_string(&self, idx: isize) -> String;
    fn to_stringx(&self, idx: isize) -> Option<String>;
    /* push functions (rust -> stack) */
    fn push_nil(&mut self);
    fn push_boolean(&mut self, b: bool);
    fn push_integer(&mut self, n: i64);
    fn push_number(&mut self, n: f64);
    fn push_string(&mut self, s: String);
    /* comparison and arithmetic functions */
    fn arith(&mut self, op: u8);
    fn compare(&self, idx1: isize, idx2: isize, op: u8) -> bool;
    /* miscellaneous functions */
    fn len(&mut self, idx: isize);
    fn concat(&mut self, n: isize);
}

File: code/csharp/ch10/luavm/api/LuaState.cs
using System;
using LuaType = System.Int32;
using ArithOp = System.Int32;
using CompareOp = System.Int32;

namespace luavm.api
{
    public delegate int CsharpFunction(LuaState luaState);

    public partial interface LuaState
    {
        int LuaType { get; set; }
        int GetTop();
        int AbsIndex(int idx);
        bool CheckStack(int n);
        void Pop(int n);
        void Copy(int fromIdx, int toIdx);
        void PushValue(int idx);
        void Replace(int idx);
        void Insert(int idx);
        void Remove(int idx);
        void Rotate(int idx, int n);
        void SetTop(int idx);

        /* access functions (stack->c#) */
        string TypeName(LuaType tp);
        LuaType Type(int idx);
        bool IsNone(int idx);
        bool IsNil(int idx);
        bool IsNoneOrNil(int idx);
        bool IsBoolean(int idx);
        bool IsInteger(int idx);
        bool IsNumber(int idx);
        bool IsString(int idx);
        bool ToBoolean(int idx);
        long ToInteger(int idx);
        (long, bool) ToIntegerX(int idx);
        double ToNumber(int idx);
        (double, bool) ToNumberX(int idx);
        string ToString(int idx);
        (string, bool) ToStringX(int idx);

        /* push functions (c# -> stack) */
        void PushNil();
        void PushBoolean(bool b);
        void PushInteger(long n);
        void PushNumber(double n);
        void PushString(string s);

        void Arith(ArithOp op);
        bool Compare(int idx1, int idx2, CompareOp op);
        void Len(int idx);
        void Concat(int n);

        void NewTable();
        void CreateTable(int nArr, int nRec);
        LuaType GetTable(int idx);
        LuaType GetField(int idx, string k);
        LuaType GetI(int idx, long i);
        void SetTable(int idx);
        void SetField(int idx, string k);
        void SetI(int idx, long n);

        int Load(ref byte[] chunk, string chunkName, string mode);
        void Call(int nArgs, int nResults);

        void PushCsharpFunction(CsharpFunction f);
        bool IsCsharpFunction(int idx);
        CsharpFunction ToCsharpFunction(int idx);

        void PushGlobalTable();
        LuaType GetGlobal(string name);
        void SetGlobal(string name);
        void Register(string name, CsharpFunction f);

        void PushCsharpClosure(CsharpFunction f, int n);
    }
}